src/Umbraco.Cms.Api.Management/Controllers/Dictionary/Tree/DictionaryTreeControllerBase.cs

```diff
@@ -54,10 +54,19 @@ protected override async Task<ActionResult<IEnumerable<NamedEntityTreeItemRespon
             }
         }
 
-        NamedEntityTreeItemResponseModel[] viewModels = await Task.WhenAll(ancestors.Select(CreateEntityTreeItemViewModelAsync));
+        NamedEntityTreeItemResponseModel[] viewModels = await MapTreeItemViewModelsAsync(ancestors).ToArrayAsync();
+
         return Ok(viewModels.Reverse());
     }
 
+    private async IAsyncEnumerable<NamedEntityTreeItemResponseModel> MapTreeItemViewModelsAsync(IEnumerable<IDictionaryItem> dictionaryItems)
```

Comment:
By using a method that returns `IAsyncEnumerable`, we're able to asynchronously iterate over the items by using `await foreach (var item in MapTreeItemViewModelsAsync(...))`, `await MapTreeItemViewModelsAsync(...).ToArrayAsync()` or any other `To...Async()` mehtods...

Reply:
Even better would be to actually return the `IAsyncEnumerable` as model (when possible), because that will avoid loading all items into memory and JSON serialize each item one by one.